### DISCUSSION TITLE
Store undefined no Component

### DISCUSSION THREAD

**Discussion by mahenrique94** [frontendbr/forum#916] [Perguntas e Respostas]
Bom dia galera, estou com um problema para implementar o **Redux** com `react-thunk`. Quando faço:

```
constructor(props) {
    super(props);
    console.log(this.props.store);
}
```

Esta imprimindo `undefined`, sendo assim, não consigo realizar o `dispatch`, o que acham que pode ser?

**Comment by felipefialho** [comment]
Já debuggou para entender o o que está vindo em `props`?

**Comment by mahenrique94** [comment]
Já, esta vindo os valores normal, apenas a store que não esta chegando.

**Comment by felipefialho** [comment]
Tente usar [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)

```
const { store } = this.props
``` 

E veja se retorna algo.

**Comment by mahenrique94** [comment]
Ja havia tentado, retorna `undefined`, abaixo segue os arquivos:

### index.js
```
import React from 'react';
import ReactDOM from 'react-dom';
import registerServiceWorker from './registerServiceWorker';
import { Provider } from "react-redux";
import { Router, Route } from "react-router";
import createHistory from "history/createBrowserHistory";
import store from "./store";

import Home from './components/home/Home';

import "./assets/css/normalize.css";
import "./assets/css/fontawesome/css/fontawesome-all.min.css";
import './index.css';

ReactDOM.render(
    <Provider store={ store }>
        <Router history={ createHistory() }>
            <Route exact path="/">
                <div className="o-app">
                    <Route component={ Home } exact path="/"/>
                </div>
            </Route>
        </Router>
    </Provider>,
    document.getElementById('root')
);

registerServiceWorker();
```

### store.js
```
import { createStore, applyMiddleware } from "redux";
import reducers from "./reducers";
import thunkMiddleware from "redux-thunk";

export default createStore(reducers, applyMiddleware(thunkMiddleware));
```

### index.js (reducers)
```
import { combineReducers } from "redux";
import { error, loading, success } from "./media";

export default combineReducers({
    error,
    loading,
    success
});
```

### Home.js (Componente ao qual não esta chegando a store)
```
import React, { Component } from "react";

import App from "../../App";
import Media from "../media/Media";

import { list } from "../../actions/media";

export default class Home extends Component {

    constructor(props) {
        super(props);
        this.state = {medias : []};
        console.log(this.props);
        console.log(this.props.store);
    }

    componentDidMount() {
        // this.props.store.dispatch(list());
    }

    render() {
        return (
            <App>
                <Media/>
            </App>
        );
    }

}
```

Com certeza é algo muito bobo, mas ainda não achei.

Caso queira ver o projeto inteiro, esta no meu github: [miniflix](https://github.com/mahenrique94/miniflix)

**Comment by felipefialho** [comment]
@mahenrique94 Coloca no [plnkr](http://plnkr.co/) pra gente ver?

**Comment by mahenrique94** [comment]
Não conheço essa ferramenta, vou dar uma olhada.

**Comment by brunolemos** [comment]
A store não é passada por props, mas sim por context.
Mas raramente você precisa usá-la assim direto, veja se o @connect não resolve o que você quer.

**Comment by mahenrique94** [comment]
Tentei por context também @brunolemos, mas o `this.context` também esta `undefined`, não preciso utilizar o connect, pois estou trabalhando com redux-thunk, ele seria necessário se fosse redux normal.

**Comment by haskellcamargo** [comment]
Jovem, `ConnectedRouter` te dá acesso à store no `react-router` versão 4.

**Comment by mahenrique94** [comment]
Já tentei também @haskellcamargo, continuou `undefined`. Eu troquei o `Router` pelo `ConnectedRouter`, é isso mesmo?

**Comment by haskellcamargo** [comment]
Ok, uma gambizinha que tu pode fazer, então, é
`<Route component={ props => <Home { ...props } store={ store } /> } exact path="/" />`, mas se perguntarem, eu não disse nada. 🤐

**Comment by mahenrique94** [comment]
Com a gambizinha deu certo @haskellcamargo, a `store` chegou na `props`, porém preciso descobrir o que esta de errado, ja tentei de todos os jeitos e nada da `store` chegar de forma normal no `context` ou `props`.

**Comment by geovanerocha** [comment]
@mahenrique94 lê isso que vai te ajudar:
http://www.thegreatcodeadventure.com/the-react-plus-redux-container-pattern/

basicamente vc faz todos os paranauês relacionados ao redux em um `${seuComponente}.container.js`, e o seu componente original fica puro sem importar connect nem nada, tudo isso fica no container.

**Comment by bernardodiasc** [comment]
Pelo que entendi o componente Home precisa do connect e `mapStateToProps` pra vc poder acessar a store.

Esse link ajuda a entender bem sobre isso https://stackoverflow.com/questions/38202572/understanding-react-redux-and-mapstatetoprops

**Comment by mahenrique94** [comment]
Obrigado @geovanerocha e @bernardodiasc, vou terminar de ler um artigo que o @haskellcamargo me passou e já vou ler o seus.

Mas eu vou chamar as actions diretas no `redux-thunk`, mesmo assim precisa fazer o `map` e `connect`? Era justamente isso que eu queria evitar.

**Comment by bernardodiasc** [comment]
O redux-thunk possui outra finalidade, ele não auto-conecta seus componentes no store, pra isso vc precisa do `connect` mesmo.

**Comment by felipefialho** [comment]
@mahenrique94 Avise se resolver :)

**Comment by mahenrique94** [comment]
Deu certo @LFeh, podem verificar se eu fiz certo:

Transformei o *component* Home em um *container*:

```
export default class Home extends Component {

    render() {
        return (
            <App>
                <Medias/>
            </App>
        );
    }

}
```

Depois criei um *component* chamado `Medias` para agrupar todas as `Media` e nele eu faço os `maps` e `connect`:

```
class Medias extends Component {

    componentDidMount() {
        this.props.list();
    }

    render() {
        return(
            <div className="o-container">
                { this.props.medias.map(media => <Media title={ media.title } slug={ media.slug }/>) }
            </div>
        );
    }

}

const mapStateToProps = state => {

    return {
        medias : state.success
    }

}

const mapDispatchToProps = dispatch => {

    return {
        list : () => dispatch(list())
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(Medias);
```

É assim mesmo?

**Comment by brunolemos** [comment]
> Tentei por context também @brunolemos, mas o this.context também esta undefined

Você não pode simplemente usar `this.context`.
Antes disso seu componente deve declarar quais contexts ele quer acessar, usando o `contextTypes`:

```jsx
class HomeContainer extends PureComponent {
  static contextTypes = {
    store: PropTypes.object,
  }

  ...
}
```

> não preciso utilizar o connect, pois estou trabalhando com redux-thunk, ele seria necessário se fosse redux normal

Acho que não tem relação uma coisa com a outra. Se vai acessar algum valor ou dispatch do redux  dentro dos seus components, recomendo usar o connect. Melhor ainda se for usando decorator (`@connect`).

@mahenrique94 Recomendo fortemente você ler a documentação do react, do redux e também ver alguns vídeos antes de cair na implementação, senão vai continuar tendo dificuldade a toa:

- https://redux.js.org/
- https://egghead.io/courses/getting-started-with-redux

**Comment by mahenrique94** [comment]
Já fiz vários cursos na Alura e School of Net sobre React e Redux, porém, ainda estava meio confuso na minha cabeça, agora que consegui e li os artigos do pessoal, deu uma clareada, até então estava confundindo algumas coisas.

Logo com a prática essa nova maneira de fluxo vai ficar mais clara ainda.
